plutus-core/plutus-ir/src/PlutusIR/Transform/Inline/Inline.hs

```diff
@@ -407,3 +432,124 @@ maybeAddTySubst tn rhs = do
         modify' (extendType tn rhs)
         pure Nothing
     else pure $ Just rhs
+
+applyAndBetaReduce ::
+  forall tyname name uni fun ann.
+  (InliningConstraints tyname name uni fun) =>
+  -- | The rhs of the variable, should have been renamed already
+  Term tyname name uni fun ann ->
+  -- | The arguments, already processed
+  AppContext tyname name uni fun ann ->
+  InlineM tyname name uni fun ann (Term tyname name uni fun ann)
+applyAndBetaReduce rhs args = do
+  info <- ask
+  -- We run the beta and inlining passes for `len args` times, since sometimes the args
+  -- can only be beta-reduced one by one. An example is
```

Comment:
I'm somewhat tempted to say that that's a limitation of the inliner that we should try to fix. I'm not sure how big a deal it is if this doesn't work as effectively for effectful args, but only running the inliner once seems attractive.

Reply:
I'd like to see how much a difference it makes if we just do it once.

Reply:
Okay, I tried it and things are much worse. However, I think this is not so much to do with effectful terms. It's to do with type arguments. We have long had the problem that you can't beta-reduce all the arguments at once for a term with type and term arguments. You end up with something like this:
```
map = (/\ a b . \(f :: a -> b) (l :: List a) . ...) {integer} {integer} incInteger ls
--->
(let
  a = integer
  b = integer
in \(f :: a -> b) (l :: List a) . ...)) incInteger ls
```
Whereas if we produced this:
```
let
  a = integer
  b = integer
  f :: a -> b = incInteger
  l :: List a = ls
in ...
```
the bindings for `f` and `l` don't typecheck, because their signatures reference the type variables, but their RHSs need the type variables to have been substituted in.

So we end up awkwardly having to do one beta, inline the type variable (which we don't always do!) and then do the rest of the arguments. This plausibly blocks a bunch of optimzation, with this being one example.

It would be good to improve this. It's not as simple as just always inlining type variables, because this is exactly the situation that we have when we compile down datatypes, and we really _don't_ want to inline all of those otherwise we get truly enormous programs and the typechecker grinds to a halt.

However, I think it _would_ be safe to always inline type variables just when doing call-site inlining, because we have the size check, so that won't let us actually grow the program!